### PR TITLE
Blogpost for kubelet credential providers

### PR DESCRIPTION
**Pull request by ndixita**
Signed-off-by: Narang Dixita Sohanlal [ndixita@google.com](https://github.com/kubernetes/website/pull/ndixita@google.com)

Adding doc changes for kep-2133 for KubeletCredentialProviders GA.
https://github.com/kubernetes/enhancements/issues/2133

### CODE REVIEW

**Review by sftim — DISMISSED** [review]
Thanks.

You need to change this text too:
```markdown
* The kubelet image credential provider is introduced in v1.20 as an alpha feature. As with other alpha features,
  a feature gate `KubeletCredentialProviders` must be enabled on only the kubelet for the feature to work.
```
as it is out of date.

---

Please also update https://kubernetes.io/docs/concepts/containers/images/#using-a-private-registry to link to https://k8s.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/ and to mark that image pull credential plugin mechanism as stable in 1.26

**Comment by sftim** [comment]
/lgtm cancel

These changes are useful, but they are not the whole set we need.

/retitle Update docs for kubelet credentials provider GA
/sig node

File: content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md [hunk 1]
```diff
@@ -7,7 +7,7 @@ description: Configure the kubelet's image credential provider plugin
 content_type: task
```
Comment (andrewsykim) [inline content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md:7]: In `content/en/docs/concepts/containers/images.md` under `## Using a private registry` add the following reference to this doc:

```
### Kubelet Credential Provider

{{< note >}}
This approach is suitable when kubelet needs to fetch registry credentials dynamically.
Most commonly used for registries provided by cloud providers where auth tokens are short-lived. 
{{< /note >}}

The Kubelet can be configured to invoke a plugin binary to dynamically fetch registry credentials for a container image.
This is the most robust and versatile way to fetch credentials for private registries, but also requires kubelet-level configuration to enable.

See the [Kubelet Credential Proivder](./content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md) docs for more details.      
```

(or something like that)

**Comment by ndixita** [comment]
@sftim I have made the recommended changes PTAL. What other changes are needed to make this as complete?

File: content/en/docs/concepts/containers/images.md [hunk 2]
```diff
@@ -180,6 +183,18 @@ For an example of configuring a private container image registry, see the
...
+
+{{< note >}}
+This approach is suitable when kubelet needs to fetch registry credentials dynamically.
+Most commonly used for registries provided by cloud providers where auth tokens are short-lived. 
+{{< /note >}}
+
+The Kubelet can be configured to invoke a plugin binary to dynamically fetch registry credentials for a container image.
+This is the most robust and versatile way to fetch credentials for private registries, but also requires kubelet-level configuration to enable.
+
+See the [Kubelet Credential Proivder](/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md) docs for more details.
```
Comment (tengqm) [inline content/en/docs/concepts/containers/images.md:196]:
```suggestion
See the [Kubelet Credential Proivder](/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/) for more details.      
```

Comment (sftim) [reply content/en/docs/concepts/containers/images.md:196]: Also Proivder is misspelled.

File: content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md [hunk 3]
```diff
@@ -54,7 +53,7 @@ should be invoked for which container images. Here's an example configuration fi
 ```yaml
 apiVersion: kubelet.config.k8s.io/v1alpha1
```
Comment (andrewsykim) [inline content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md:54]: Use v1

Comment (andrewsykim) [reply content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md:54]: Same on line 101

File: content/en/docs/concepts/containers/images.md [hunk 4]
```diff
@@ -180,6 +183,18 @@ For an example of configuring a private container image registry, see the
 [Pull an Image from a Private Registry](/docs/tasks/configure-pod-container/pull-image-private-registry)
 task. That example uses a private registry in Docker Hub.
 
+### Kubelet Credential Provider
```
Comment (sftim) [inline content/en/docs/concepts/containers/images.md:186]: (nit)
```suggestion
### Kubelet credential provider for authenticated image pulls {#kubelet-credential-provider}
```

File: content/en/docs/concepts/containers/images.md [hunk 5]
```diff
@@ -180,6 +183,18 @@ For an example of configuring a private container image registry, see the
...
+
+{{< note >}}
+This approach is suitable when kubelet needs to fetch registry credentials dynamically.
```
Comment (sftim) [inline content/en/docs/concepts/containers/images.md:189]:
```suggestion
This approach is especially suitable when kubelet needs to fetch registry credentials dynamically.
```
You can use these plugins to give the kubelet access to long lived credentials.

File: content/en/docs/concepts/containers/images.md [hunk 6]
```diff
@@ -180,6 +183,18 @@ For an example of configuring a private container image registry, see the
...
+Most commonly used for registries provided by cloud providers where auth tokens are short-lived. 
+{{< /note >}}
+
+The Kubelet can be configured to invoke a plugin binary to dynamically fetch registry credentials for a container image.
```
Comment (sftim) [inline content/en/docs/concepts/containers/images.md:193]:
```suggestion
You can configure the kubelet to invoke a plugin binary to dynamically fetch registry credentials for a container image.
```

**Review by sftim — COMMENTED** [review]
I noticed another thing. The file `content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md` should really be
`content/en/docs/tasks/kubelet-credential-provider.md` and we should configure a redirect (see `static/_redirects`) for the old URL.

---

I added inline feedback.

File: content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md [hunk 7]
```diff
@@ -27,8 +27,7 @@ This guide demonstrates how to configure the kubelet's image credential provider
 
 ## {{% heading "prerequisites" %}}
 
-* The kubelet image credential provider is introduced in v1.20 as an alpha feature. As with other alpha features,
-  a feature gate `KubeletCredentialProviders` must be enabled on only the kubelet for the feature to work.
+* The kubelet image credential provider has been a beta feature since v1.24.0 and is enabled by default. With v1.26, the feature has reached General Availability (GA).
```
Comment (sftim) [inline content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md:30]:
```suggestion
* You need a Kubernetes cluster with nodes that support kubelet credential
  provider plugins. This support is available in Kubernetes {{< skew currentVersion >}};
  Kubernetes v1.24 and v1.25 included this as a beta feature, enabled by default.
```

The feature state shortcode, above, shows when the feature reached GA.

File: content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md [hunk 8]
```diff
@@ -27,8 +27,7 @@ This guide demonstrates how to configure the kubelet's image credential provider
...
 * A working implementation of a credential provider exec plugin. You can build your own plugin or use one provided by cloud providers.
```
Comment (sftim) [inline content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md:32]:
```suggestion

{{< version-check >}}

```
This tells readers how to find out what version of Kubernetes they have.

**Comment by ndixita** [comment]
> I noticed another thing. The file `content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md` should really be `content/en/docs/tasks/kubelet-credential-provider.md` and we should configure a redirect (see `static/_redirects`) for the old URL.
> 
@sftim I don't see any readmes directly in tasks folder. Everything has a dedicated folder. Is my understanding correct?

**Comment by sftim** [comment]
> I don't see any readmes directly in tasks folder. Everything has a dedicated folder. Is my understanding correct?

Ah, good point. How about putting this task inside https://kubernetes.io/docs/tasks/administer-cluster/?

**Comment by leonardpahlke** [comment]
@ndixita How much work is needed on this one? I would love to see this get in asap to make it before releasing K8s 1.26 in less than a week.

cc @krol3

**Comment by ndixita** [comment]
@sftim @andrewsykim do we also need content/en/docs/reference/config-api/kubelet-credentialprovider.v1beta1.md for v1? 

There are references to v1alpha1 config in here 
https://kubernetes.io/docs/tasks/kubelet-credential-provider/kubelet-credential-provider/#what-s-next

**Comment by andrewsykim** [comment]
@ndixita I think we should update all references to v1 only

**Comment by krol3** [comment]
@ndixita  is this PR https://github.com/kubernetes/website/pull/38006 valid for this release 1.26?

**Comment by sftim** [comment]
We often use tabs to document that an old API is also still supported (showing the stable API by default). Is the v1 API different from v1beta1 for kubelet image pull credential plugins?

**Comment by ndixita** [comment]
> We often use tabs to document that an old API is also still supported (showing the stable API by default). Is the v1 API different from v1beta1 for kubelet image pull credential plugins?

No, the v1 is same as v1beta1

File: content/en/docs/tasks/administer-cluster/kubelet-credential-provider.md [hunk 9]
```diff
@@ -27,10 +27,13 @@ This guide demonstrates how to configure the kubelet's image credential provider
 
 ## {{% heading "prerequisites" %}}
 
-* The kubelet image credential provider is introduced in v1.20 as an alpha feature. As with other alpha features,
-  a feature gate `KubeletCredentialProviders` must be enabled on only the kubelet for the feature to work.
+* You need a Kubernetes cluster with nodes that support kubelet credential
+  provider plugins. This support is available in Kubernetes {{< skew currentVersion >}};
```
[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubelet-credential-provider.md:31]: The aim of `{{< skew currentVersion >}}` is to make it clear that **future** Kubernetes versions do support this feature.
I specifically wanted readers to see that.

IMO what we ought to do as well is to set `min-kubernetes-server-version:` in the front matter (try grepping other pages to see an example) and mark it as v1.26
If we include that, the final rendered page should make the situation nicely clear to readers.

[the remark this answers is not in the document]
Comment (sftim) [reply content/en/docs/tasks/administer-cluster/kubelet-credential-provider.md:31]: (if the shortcodes we have are not easy enough to understand, I advocate treating that as a defect and tracking an issue about it)

File: content/en/docs/concepts/containers/images.md [hunk 10]
```diff
@@ -180,6 +183,18 @@ For an example of configuring a private container image registry, see the
...
+
+{{< note >}}
+This approach is especially suitable when kubelet needs to fetch registry credentials dynamically.
+Most commonly used for registries provided by cloud providers where auth tokens are short-lived. 
+{{< /note >}}
+
+You can configure the kubelet to invoke a plugin binary to dynamically fetch registry credentials for a container image.
+This is the most robust and versatile way to fetch credentials for private registries, but also requires kubelet-level configuration to enable.
+
+See the [Kubelet Credential Provider](/docs/tasks/administer-cluster/kubelet-credential-provider/) for more details.
```
Comment (sftim) [inline content/en/docs/concepts/containers/images.md:196]: (nit)
```suggestion
See [Configure a kubelet image credential provider](/docs/tasks/administer-cluster/kubelet-credential-provider/) for more details.
```

**Comment by krol3** [comment]
@ndixita Please could you review the suggestions changes? We need to close this PR before the 1.26 release.
cc: @leonardpahlke @reylejano

**Comment by derekwaynecarr** [comment]
the technical content is lgtm for me for sig-node.

/lgtm

File: content/en/docs/tasks/administer-cluster/kubelet-credential-provider.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (ndixita) [reply content/en/docs/tasks/administer-cluster/kubelet-credential-provider.md:32]: Added the min-kubernetes-server-version: v1.26

**Comment by sftim** [comment]
@tengqm, are you happy with how this previews now?

**Comment by sftim** [comment]
@ndixita: when you rebase this, would you be willing to squash it down to 1 or 2 commits as well?

**Comment by tengqm** [comment]
/lgtm
/approve
Thanks.

**Pull request merged**